system/doc/reference_manual/opaques.xml

```diff
@@ -0,0 +1,141 @@
+
+    </legalnotice>
+
+    <title>Opaques</title>
+    <prepared>Max Heiber</prepared>
+    <docno></docno>
+    <date></date>
+    <rev></rev>
+    <file>opaques.xml</file>
+  </header>
+
+  <section>
+    <title>Opaque Type Aliases</title>
+    <p>The main use case for opacity in Erlang is to hide the implementation of a data type, enabling evolving the API while minimizing the risk of breaking consumers. The runtime does not check opacity. Dialyzer provides some opacity-checking, but the rest is up to convention.
+    </p>
+    <p>
+      This document explains what Erlang opacity is (and the trade-offs involved) via the example of OTP's
+      <c>sets:set()</c>
+      data type. This type
+      <em>was</em>
+      defined in `sets` module like this:
+    </p>
+    <code type="erl">-opaque set(Element) :: #set{segs :: segs(Element)}.</code>
+    <p>OTP 24  changed the definition to the following, in
+    <url href="https://github.com/erlang/otp/commit/e66941e8d7c47b973dff94c0308ea85a6be1958e">this commit</url>
+    </p>
+    <code type="erl">-opaque set(Element) :: #set{segs :: segs(Element)} | #{Element => ?VALUE}.</code>
+    <p>
+      And this change was safer and more backwards-compatible than if the type had been defined with
+      <c>-type</c>
+      instead of
+      <c>-opaque</c>
+      . Here's why: when a module defines an
+      <c>-opaque</c>
+      , the contract is that only the defining module should rely on the definition of the type: no other modules should rely on the definition.
+    </p>
+    <p>
+      This means that code that pattern-matched on
+      <c>set</c>
+      as a record/tuple technically broke the contract, and opted in to being potentially broken when the definition of
+      <c>set()</c>
+      changed. Before OTP 24, this code printed
+      <c>ok</c>
+      . In OTP 24 it may error:
+    </p>
+    <code type="erl">
+case sets:new() of
+    Set when is_tuple(Set) -&gt;
+        io:format(&quot;ok&quot;)
+end.
+    </code>
+    <p>
+      <strong>When working with an opaque defined in another module, here are some recommendations:</strong>
+    </p>
+
+    <list type="bulleted">
+      <item>
+        Don't examine the underlying type using pattern-matching, guards, or functions that reveal the type, such as
+        <c>tuple_size/1</c>
+        .
+      </item>
+      <item>
+        Instead, use functions provided by the module for working with the type. For example,
+        <c>sets</c>
+        module provides
+        <c>sets:new/0</c>
+        ,
+        <c>sets:add/2</c>
+        ,
+        <c>sets:is_element/2</c>
+        , etc.
+      </item>
+      <item>
+        <c>sets:set(a)</c>
+        is a subtype of
+        <c>sets:set(a | b)</c>
+        and not the other way around. Generally, you can rely on the property that
+        <c>the_opaque(T)</c>
+        is a subtype of
+        <c>the_opaque(U)</c>
+        when T is a subtype of U.
+      </item>
+    </list>
```

Comment:
Another interesting pattern is when one stores opaque type in, eg, optional record field (or generally in some container) and then have some code that checks if record's field have been initialized:

```erlang
-record(lazy_set, {the_set :: sets:set() | undefined, ...}).

add(Val, #lazy_set{the_set = undefined} = LS) ->
   add(Val, add(#lazy_set{the_set = sets:new()}));
add(Val, #lazy_set{the_set = Set)) ->
  ...
```
or similar with maps
```erlang
-type my_type() :: #{my_key => sets:set()}.

-spec my_fun(..., my_type()) -> ...
my_fun(..., Map) ->
  case maps:get(my_key, Map, undefined) of
    undefined ->  ...;
    Set -> ...
  end
```

So, the recomendation, I believe, is to wrap the opaque value to some enum-style tuple, or have some flag in a separate field

```
-record(lazy_set, {the_set :: {set, sets:set()} | undefined}).
```